intermediate_source/parametrizations.py

```diff
@@ -0,0 +1,353 @@
+# Intializing Parametrizations
+# ----------------------------
+# Parametrizations come with a mechanism to initialize them. If we implement a method
+# ``right_inverse`` with signature
+#
+# .. code-block:: python
+#
+#     def right_inverse(self, X: Tensor) -> Tensor
+#
+# it will be used when assigning to the parametrized tensor.
+#
+# Let's upgrade our implementation of the ``Skew`` class to support this
+class Skew(nn.Module):
+    def forward(self, X):
+        A = X.triu(1)
+        return A - A.transpose(-1, -2)
+
+    def is_skew(self, X):
+        return torch.allclose(X, -X.transpose(-1, -2))
+
+    def right_inverse(self, A):
+        if not self.is_skew(A):
+            raise ValueError(f"The provided matrix {A} is not skew-symmetric")
+        return A.triu(1)
+
+###############################################################################
+# We may now initialize a layer that is parametrized with ``Skew``
+layer = nn.Linear(3, 3)
+P.register_parametrization(layer, "weight", Skew())
+X = torch.rand(3, 3)
+X = X - X.T                             # X is now skew-symmetric
+layer.weight = X                        # Initialize layer.weight to be X
+assert torch.allclose(layer.weight, X)  # layer.weight == X
+
+###############################################################################
+# This ``right_inverse`` works as expected when we compose parametrizations. To see this, let's
+# upgrade the Cayley parametrization to also support being initialized
+class CayleyMap(nn.Module):
+    def __init__(self, n):
+        self.register_buffer("Id", torch.eye(n))
+
+    def forward(self, X):
+        # (I + X)(I - X)^{-1}
+        return torch.solve(self.Id + X, self.Id - X).solution
+
+    def right_inverse(self, A):
+        # See https://en.wikipedia.org/wiki/Cayley_transform#Matrix_map
+        # (X - I)(X + I)^{-1}
+        return torch.solve(X - self.Id, self.Id + X).solution
+
+layer_orthogonal = nn.Linear(3, 3)
+P.register_parametrization(layer_orthogonal, "weight", Skew())
+P.register_parametrization(layer_orthogonal, "weight", CayleyMap(3))
+# Sample an orthogonal matrix with positive determinant
+X = torch.empty(3, 3)
+nn.init.orthogonal_(X)
+if X.det() < 0.:
+    X[0].neg_()
+layer_orthogonal.weight = X
+assert torch.allclose(X, layer_orthogonal.weight)  # layer_orthogonal.weight == X
+
+###############################################################################
+# This initialization step can be written more succinctly as
+layer_orthogonal.weight = nn.init.orthogonal_(layer_orthogonal.weight)
+
+###############################################################################
+# The name of this method comes from the fact that we would often expect
+# that ``forward(right_inverse(X)) == X``. This is a direct way of rewriting that
+# the forward afer the initalization with value ``X`` should return the value ``X``.
+# This constraint is not enforced in the code. In fact, at times, it might be of
```

Comment:
Maybe "This constraint is not enforced in the code" -> "This constraint is not strongly enforced in practice"